File: mci/parser.py
from __future__ import annotations

import json
from typing import Dict, List, Optional

from .types import (
    CategoryInput,
    CourseInput,
    CurriculumFile,
    CurriculumInput,
    RequirementInput,
)

# ─────────────────────────────────────────────────────────────────────────────
# Parser / validação
# ─────────────────────────────────────────────────────────────────────────────


class ParseError(Exception):
    pass


def parse(raw_json: str) -> CurriculumFile:
    try:
        data = json.loads(raw_json)
    except json.JSONDecodeError as e:
        raise ParseError(f"O arquivo não contém um JSON válido: {e}")

    if not isinstance(data, dict):
        raise ParseError("O JSON de entrada deve ser um objeto.")

    curriculum = _parse_curriculum(data.get("curriculum"))
    courses = _parse_courses(data.get("courses"))
    requirements = _parse_requirements(data.get("requirements"))
    categories = _parse_categories(data.get("categories"))
    card_fill_style = _parse_display(data.get("display"))

    _check_course_category_references(courses, categories)

    return CurriculumFile(
        curriculum=curriculum,
        courses=courses,
        requirements=requirements,
        categories=categories,
        card_fill_style=card_fill_style,
    )


def _parse_curriculum(value) -> CurriculumInput:
    if not isinstance(value, dict):
        raise ParseError('Campo "curriculum" ausente ou inválido.')
    for f in ("code", "name", "availableSince", "description"):
        if not isinstance(value.get(f), str) or not value[f].strip():
            raise ParseError(f'Campo "curriculum.{f}" ausente ou vazio.')
    if not isinstance(value.get("levels"), (int, float)) or value["levels"] < 1:
        raise ParseError('Campo "curriculum.levels" deve ser um número positivo.')
    return CurriculumInput(
        code=value["code"],
        name=value["name"],
        available_since=value["availableSince"],
        description=value["description"],
        levels=int(value["levels"]),
    )


def _parse_courses(value) -> List[CourseInput]:
    if not isinstance(value, list):
        raise ParseError('Campo "courses" deve ser um array.')
    courses = []
    for i, item in enumerate(value):
        if not isinstance(item, dict):
            raise ParseError(f"courses[{i}]: item inválido.")
        for f in ("code", "name", "syllabus"):
            if not isinstance(item.get(f), str) or not item[f].strip():
                raise ParseError(f"courses[{i}].{f}: ausente ou vazio.")
        for f in ("hours", "credits", "level"):
            if not isinstance(item.get(f), (int, float)) or item[f] < 1:
                raise ParseError(f"courses[{i}].{f}: deve ser um número positivo.")
        if not isinstance(item.get("tags"), list):
            raise ParseError(f"courses[{i}].tags: deve ser um array.")
        for j, tag in enumerate(item["tags"]):
            if not isinstance(tag, str) or not tag.strip():
                raise ParseError(f"courses[{i}].tags[{j}]: deve ser string não vazia.")
        cat = item.get("category")
        if cat is not None and (not isinstance(cat, str) or not cat.strip()):
            raise ParseError(
                f"courses[{i}].category: deve ser string não vazia, quando informado."
            )
        courses.append(
            CourseInput(
                code=item["code"],
                name=item["name"],
                hours=int(item["hours"]),
                credits=int(item["credits"]),
                level=int(item["level"]),
                syllabus=item["syllabus"],
                tags=list(item["tags"]),
                category=cat,
            )
        )
    return courses


def _parse_categories(value) -> List[CategoryInput]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise ParseError('Campo "categories" deve ser um array, quando informado.')
    categories = []
    for i, item in enumerate(value):
        if not isinstance(item, dict):
            raise ParseError(f"categories[{i}]: item inválido.")
        if not isinstance(item.get("id"), str) or not item["id"].strip():
            raise ParseError(f"categories[{i}].id: ausente ou vazio.")
        if not isinstance(item.get("name"), str) or not item["name"].strip():
            raise ParseError(f"categories[{i}].name: ausente ou vazio.")
        color = item.get("color")
        if color is not None and (not isinstance(color, str) or not color.strip()):
            raise ParseError(
                f"categories[{i}].color: deve ser string não vazia, quando informado."
            )
        categories.append(CategoryInput(id=item["id"], name=item["name"], color=color))
    return categories


def _parse_display(value) -> Optional[str]:
    if value is None:
        return None
    if not isinstance(value, dict):
        raise ParseError('Campo "display" deve ser um objeto, quando informado.')
    fill_style = value.get("card_fill_style")
    if fill_style is not None and fill_style != "category":
        raise ParseError(
            'Campo "display.card_fill_style" deve ser "category" quando informado.'
        )
    return fill_style


def _parse_requirements(value) -> List[RequirementInput]:
    if not isinstance(value, list):
        raise ParseError('Campo "requirements" deve ser um array.')
    valid_types = {"prerequisite", "special", "corequisite", "credit_requirement"}
    reqs = []
    for i, item in enumerate(value):
        if not isinstance(item, dict):
            raise ParseError(f"requirements[{i}]: item inválido.")
        rtype = item.get("type")
        if rtype not in valid_types:
            raise ParseError(f'requirements[{i}].type: valor inválido "{rtype}".')
        to = item.get("to")
        if not isinstance(to, str) or not to.strip():
            raise ParseError(f"requirements[{i}].to: ausente ou vazio.")
        from_code = item.get("from")
        if rtype != "credit_requirement":
            if not isinstance(from_code, str) or not from_code.strip():
                raise ParseError(f"requirements[{i}].from: ausente ou vazio.")
        if rtype == "credit_requirement":
            mc = item.get("min_credits")
            if not isinstance(mc, (int, float)) or mc < 1:
                raise ParseError(
                    f"requirements[{i}].min_credits: deve ser um número positivo."
                )
        reqs.append(
            RequirementInput(
                type=rtype,
                to=to,
                from_code=from_code if isinstance(from_code, str) else None,
                description=item.get("description"),
                min_credits=(
                    int(item["min_credits"])
                    if item.get("min_credits") is not None
                    else None
                ),
            )
        )
    return reqs


def _check_course_category_references(
    courses: List[CourseInput], categories: List[CategoryInput]
) -> None:
    cat_ids = {c.id for c in categories}
    for course in courses:
        if course.category and course.category not in cat_ids:
            raise ParseError(
                f'courses[{course.code}].category refere-se a categoria inexistente: "{course.category}".'
            )

File: src/parser.ts
import { CurriculumFile, CourseInput, RequirementInput } from './types';

// ─── Erros de validação ───────────────────────────────────────────────────────

export class ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ParseError';
  }
}

// ─── Ponto de entrada ─────────────────────────────────────────────────────────

export function parse(rawJson: string): CurriculumFile {
  let data: unknown;
  try {
    data = JSON.parse(rawJson);
  } catch {
    throw new ParseError('O arquivo não contém um JSON válido.');
  }

  assertCurriculumFile(data);
  validate(data);
  return data;
}

// ─── Asserções de estrutura ──────────────────────────────────────────────────

function assertCurriculumFile(data: unknown): asserts data is CurriculumFile {
  if (typeof data !== 'object' || data === null) {
    throw new ParseError('O JSON de entrada deve ser um objeto.');
  }

  const d = data as Record<string, unknown>;

  assertCurriculum(d['curriculum']);
  assertCourses(d['courses']);
  assertRequirements(d['requirements']);
  assertCategories(d['categories']);
  assertDisplay(d['display']);
}

function assertCurriculum(value: unknown): void {
  if (typeof value !== 'object' || value === null) {
    throw new ParseError('Campo "curriculum" ausente ou inválido.');
  }
  const c = value as Record<string, unknown>;
  for (const field of ['code', 'name', 'availableSince', 'description']) {
    if (typeof c[field] !== 'string' || (c[field] as string).trim() === '') {
      throw new ParseError(`Campo "curriculum.${field}" ausente ou vazio.`);
    }
  }
  if (typeof c['levels'] !== 'number' || c['levels'] < 1) {
    throw new ParseError('Campo "curriculum.levels" deve ser um número positivo.');
  }
}

function assertCourses(value: unknown): void {
  if (!Array.isArray(value)) {
    throw new ParseError('Campo "courses" deve ser um array.');
  }
  value.forEach((item, i) => {
    const c = item as Record<string, unknown>;
    for (const field of ['code', 'name', 'syllabus']) {
      if (typeof c[field] !== 'string' || (c[field] as string).trim() === '') {
        throw new ParseError(`courses[${i}].${field}: ausente ou vazio.`);
      }
    }
    for (const field of ['hours', 'credits', 'level']) {
      if (typeof c[field] !== 'number' || (c[field] as number) < 1) {
        throw new ParseError(`courses[${i}].${field}: deve ser um número positivo.`);
      }
    }
    if (!Array.isArray(c['tags'])) {
      throw new ParseError(`courses[${i}].tags: deve ser um array.`);
    }
    (c['tags'] as unknown[]).forEach((tag, j) => {
      if (typeof tag !== 'string' || tag.trim() === '') {
        throw new ParseError(`courses[${i}].tags[${j}]: deve ser string não vazia.`);
      }
    });

    if (c['category'] !== undefined) {
      if (typeof c['category'] !== 'string' || (c['category'] as string).trim() === '') {
        throw new ParseError(`courses[${i}].category: deve ser string não vazia, quando informado.`);
      }
    }
  });
}

function assertCategories(value: unknown): void {
  if (value === undefined) return;
  if (!Array.isArray(value)) {
    throw new ParseError('Campo "categories" deve ser um array, quando informado.');
  }

  value.forEach((item, i) => {
    if (typeof item !== 'object' || item === null) {
      throw new ParseError(`categories[${i}]: item inválido.`);
    }
    const c = item as Record<string, unknown>;
    if (typeof c['id'] !== 'string' || (c['id'] as string).trim() === '') {
      throw new ParseError(`categories[${i}].id: ausente ou vazio.`);
    }
    if (typeof c['name'] !== 'string' || (c['name'] as string).trim() === '') {
      throw new ParseError(`categories[${i}].name: ausente ou vazio.`);
    }
    if (c['color'] !== undefined && (typeof c['color'] !== 'string' || (c['color'] as string).trim() === '')) {
      throw new ParseError(`categories[${i}].color: deve ser string não vazia, quando informado.`);
    }
  });
}

function assertDisplay(value: unknown): void {
  if (value === undefined) return;
  if (typeof value !== 'object' || value === null) {
    throw new ParseError('Campo "display" deve ser um objeto, quando informado.');
  }

  const d = value as Record<string, unknown>;
  const fillStyle = d['card_fill_style'];
  if (fillStyle !== undefined && fillStyle !== 'category') {
    throw new ParseError('Campo "display.card_fill_style" deve ser "category" quando informado.');
  }
}

function assertRequirements(value: unknown): void {
  if (!Array.isArray(value)) {
    throw new ParseError('Campo "requirements" deve ser um array.');
  }
  const validTypes = ['prerequisite', 'special', 'corequisite', 'credit_requirement'];
  value.forEach((item, i) => {
    const r = item as Record<string, unknown>;
    if (!validTypes.includes(r['type'] as string)) {
      throw new ParseError(`requirements[${i}].type: valor inválido "${r['type']}".`);
    }
    if (typeof r['to'] !== 'string' || (r['to'] as string).trim() === '') {
      throw new ParseError(`requirements[${i}].to: ausente ou vazio.`);
    }
    if (r['type'] !== 'credit_requirement') {
      if (typeof r['from'] !== 'string' || (r['from'] as string).trim() === '') {
        throw new ParseError(`requirements[${i}].from: ausente ou vazio.`);
      }
    }
    if (r['type'] === 'credit_requirement') {
      if (typeof r['min_credits'] !== 'number' || (r['min_credits'] as number) < 1) {
        throw new ParseError(`requirements[${i}].min_credits: deve ser um número positivo.`);
      }
    }
  });
}

// ─── Regras de negócio ────────────────────────────────────────────────────────

function validate(data: CurriculumFile): void {
  const courseMap = buildCourseMap(data.courses);
  const categoryIdSet = new Set((data.categories ?? []).map(c => c.id));

  checkDuplicateCodes(data.courses);
  checkDuplicateCategoryIds(data.categories ?? []);
  checkCourseCategoryReferences(data.courses, categoryIdSet);
  checkRequirementReferences(data.requirements, courseMap);
  checkPrerequisiteLevels(data.requirements, courseMap);
}

function buildCourseMap(courses: CourseInput[]): Map<string, CourseInput> {
  const map = new Map<string, CourseInput>();
  for (const course of courses) {
    map.set(course.code, course);
  }
  return map;
}

function checkDuplicateCodes(courses: CourseInput[]): void {
  const seen = new Set<string>();
  for (const course of courses) {
    if (seen.has(course.code)) {
      throw new ParseError(`Código de disciplina duplicado: "${course.code}".`);
    }
    seen.add(course.code);
  }
}

function checkDuplicateCategoryIds(categories: Array<{ id: string }>): void {
  const seen = new Set<string>();
  for (const category of categories) {
    if (seen.has(category.id)) {
      throw new ParseError(`Categoria duplicada em "categories": "${category.id}".`);
    }
    seen.add(category.id);
  }
}

function checkCourseCategoryReferences(courses: CourseInput[], categoryIds: Set<string>): void {
  for (const course of courses) {
    if (!course.category) continue;
    if (!categoryIds.has(course.category)) {
      throw new ParseError(
        `Disciplina "${course.code}": category "${course.category}" não existe em "categories".`
      );
    }
  }
}

function checkRequirementReferences(
  requirements: RequirementInput[],
  courseMap: Map<string, CourseInput>
): void {
  for (const req of requirements) {
    if (!courseMap.has(req.to)) {
      throw new ParseError(`Requisito: código de disciplina desconhecido em "to": "${req.to}".`);
    }
    if (req.from !== undefined && !courseMap.has(req.from)) {
      throw new ParseError(`Requisito: código de disciplina desconhecido em "from": "${req.from}".`);
    }
  }
}

function checkPrerequisiteLevels(
  requirements: RequirementInput[],
  courseMap: Map<string, CourseInput>
): void {
  for (const req of requirements) {
    if (req.from === undefined) continue;

    const from = courseMap.get(req.from)!;
    const to = courseMap.get(req.to)!;

    if (req.type === 'corequisite') {
      if (from.level !== to.level) {
        throw new ParseError(
          `Co-requisito inválido: "${req.from}" (nível ${from.level}) e "${req.to}" (nível ${to.level}) devem estar no mesmo nível.`
        );
      }
    }

    if (req.type === 'prerequisite' || req.type === 'special') {
      if (from.level >= to.level) {
        throw new ParseError(
          `Pré-requisito inválido: "${req.from}" (nível ${from.level}) deve ser de nível anterior a "${req.to}" (nível ${to.level}).`
        );
      }
    }
  }
}
